--- a/configure	2016-11-20 10:18:36.000000000 -0500
+++ b/configure	2017-02-11 15:01:21.315653217 -0500
@@ -5625,7 +5625,7 @@
 echo "$as_me: error: Cannot find libtool" >&2;}
    { (exit 1); exit 1; }; }
 	fi
-	LIB_CREATE='${LIBTOOL} --mode=link ${CC} -rpath ${DESTDIR}${libdir} ${LIBTOOL_VERSION} `cut -f1 ${top_srcdir}/VERSION` ${LIBTOOL_OPTS} ${LT_UNDEF} $(LIBS) -o'
+	LIB_CREATE='${LIBTOOL} --mode=link ${CC} -rpath ${DESTDIR}${libdir} ${LIBTOOL_VERSION} `cut -f1 ${top_srcdir}/VERSION` ${LDFLAGS} ${LIBTOOL_OPTS} ${LT_UNDEF} $(LIBS) -o'
 	LIB_OBJECT='${OBJECTS:.o=.lo}'
 	LIB_SUFFIX=.la
 	LIB_CLEAN='${LIBTOOL} --mode=clean'
--- a/menu.c	2016-11-20 13:56:05.000000000 -0500
+++ b/menu.c	2017-02-11 15:05:32.227412552 -0500
@@ -61,8 +61,22 @@
       /* *INDENT-EQLS* */
       int x1   = (menuLocation[x] == LEFT) ? x : rightcount--;
       int x2;
-      int y1   = (menuPos == BOTTOM) ? (ymax - 1) : 0;
-      int y2   = (menuPos == BOTTOM) ? (ymax - subsize[x] - 2) : TITLELINES;
+      int y1, y2;
+      if (menuPos == BOTTOM)
+      {
+         y1 = ymax - 1;
+         y2 = ymax - subsize[x] - 2;
+      }
+      else if (menuPos == TOP)
+      {
+         y1 = 0;
+         y2 = TITLELINES;
+      }
+      else
+      {
+         y1 = menuPos;
+         y2 = menuPos + TITLELINES;
+      }
       int high = subsize[x] + TITLELINES;
       int max  = -1;
 
