--- a/ui/helpers.c	2017-01-09 08:41:02.000000000 -0500
+++ b/ui/helpers.c	2017-04-20 14:02:11.163668438 -0400
@@ -75,7 +75,8 @@
 
 	char *bitmap = malloc(5 * sizeof(char));
 	bitmap[4] = '\0';
-	for(int i = 3; i >= 0; i--) {
+	int i;
+	for(i = 3; i >= 0; i--) {
 		bitmap[i] = digit % 2 ? '1' : '0';
 		digit /= 2;
 	}
--- a/ui/irqbalance-ui.c	2017-01-09 08:41:02.000000000 -0500
+++ b/ui/irqbalance-ui.c	2017-04-20 14:03:38.574977353 -0400
@@ -145,9 +145,10 @@
 
 	if(strncmp(token, "BANNED", strlen("BANNED"))) goto out;
 	token = strtok_r(ptr, " ", &ptr);
-	for(int i = strlen(token) - 1; i >= 0; i--) {
+	int i, j;
+	for(i = strlen(token) - 1; i >= 0; i--) {
 		char *map = hex_to_bitmap(token[i]);
-		for(int j = 3; j >= 0; j--) {
+		for(j = 3; j >= 0; j--) {
 			if(map[j] == '1') {
 				uint64_t *banned_cpu = malloc(sizeof(uint64_t));
 				*banned_cpu = (4 * (strlen(token) - (i + 1)) + (4 - (j + 1)));
--- a/ui/ui.c	2017-01-09 08:41:02.000000000 -0500
+++ b/ui/ui.c	2017-04-20 14:04:30.556566379 -0400
@@ -25,7 +25,8 @@
 		snprintf(top + strlen(top), COLS - strlen(top), " ");
 	}
 	mvprintw(0, 0, top);
-	for(int i = 0; i < LINES; i++) {
+	int i;
+	for(i = 0; i < LINES; i++) {
 		mvprintw(i, 0, " ");
 		mvprintw(i, COLS - 1, " ");
 	}
@@ -674,7 +675,8 @@
 	char *spaces = "    \0";
 	char indent[32] = "\0";
 	char *asciitree = " `--\0";
-	for(int i = node->type; i <= OBJ_TYPE_NODE; i++) {
+	int i;
+	for(i = node->type; i <= OBJ_TYPE_NODE; i++) {
 		snprintf(indent + strlen(indent), 32 - strlen(indent), "%s", spaces);
 		if(i != OBJ_TYPE_NODE) {
 			snprintf(indent + strlen(indent), 32 - strlen(indent), "   ");
