--- a/udev-md-raid-assembly.rules	2017-01-09 16:46:34.000000000 -0500
+++ b/udev-md-raid-assembly.rules	2018-10-11 21:51:57.806607996 -0400
@@ -5,6 +5,10 @@
 ENV{ANACONDA}=="?*", GOTO="md_inc_end"
 # assemble md arrays
 
+# If 'raid=noautodetect' is set, we shouldn't assemble MD arrays here either
+IMPORT{cmdline}="raid"
+ENV{raid}=="noautodetect", GOTO="md_inc_end"
+
 SUBSYSTEM!="block", GOTO="md_inc_end"
 
 # handle potential components of arrays (the ones supported by md)
