--- a/setup.py	2013-06-14 09:34:50.330076551 -0400
+++ b/setup.py	2013-06-14 10:02:08.751921349 -0400
@@ -278,6 +278,11 @@
                           (ext.name, sys.exc_info()[1]))
             self.failed.append(ext.name)
             return
+
+        # Don't check imports
+        self.announce('WARNING: skipping import check for "%s" (ESOS patch)' % ext.name)
+        return
+
         # Workaround for Mac OS X: The Carbon-based modules cannot be
         # reliably imported into a command-line Python
         if 'Carbon' in ext.extra_link_args:
--- a/setup.py	2014-01-22 21:52:25.725510048 -0500
+++ b/setup.py	2014-01-22 21:53:01.539274302 -0500
@@ -642,8 +642,8 @@
         # use the same library for the readline and curses modules.
         if 'curses' in readline_termcap_library:
             curses_library = readline_termcap_library
-        elif self.compiler.find_library_file(lib_dirs, 'ncursesw'):
-            curses_library = 'ncursesw'
+#        elif self.compiler.find_library_file(lib_dirs, 'ncursesw'):
+#            curses_library = 'ncursesw'
         elif self.compiler.find_library_file(lib_dirs, 'ncurses'):
             curses_library = 'ncurses'
         elif self.compiler.find_library_file(lib_dirs, 'curses'):
--- a/setup.py	2015-08-19 16:51:39.364687224 -0400
+++ b/setup.py	2015-08-19 16:52:16.832361428 -0400
@@ -1096,7 +1096,6 @@
                                   include_dirs=["Modules/_sqlite",
                                                 sqlite_incdir],
                                   library_dirs=sqlite_libdir,
-                                  runtime_library_dirs=sqlite_libdir,
                                   extra_link_args=sqlite_extra_link_args,
                                   libraries=["sqlite3",]))
         else:
