--- a/heartbeat/Raid1	2022-08-25 18:04:08.883013203 -0400
+++ b/heartbeat/Raid1	2022-08-25 18:03:56.440967830 -0400
@@ -182,7 +182,13 @@
 
 md_assemble() {
 	local mddev=$1
-	$MDADM --assemble $mddev --config=$RAIDCONF $MDADM_HOMEHOST
+	ocf_run $MDADM --assemble --run $mddev --verbose --config=$RAIDCONF $MDADM_HOMEHOST
+        rc=$?
+        if [ ${rc} -eq 1 ]; then
+           ocf_log info "Initial assembly of ${mddev} failed.  Retrying with --force."
+           /opt/quantum/scripts/tickets.sh note "Initial assembly of ${mddev} failed.  Retrying with force mode.  Potential data loss."
+           ocf_run $MDADM --assemble --run $mddev --verbose --config=$RAIDCONF --force $MDADM_HOMEHOST
+        fi
 	udev_settle --exit-if-exists=$mddev
 }
 #
@@ -226,7 +232,7 @@
 	local mddev=$1
 	local rc
 	ocf_log info "Attempting to mark array $mddev readonly"
-	$MDADM --readonly $mddev --config=$RAIDCONF
+	ocf_run $MDADM --readonly $mddev --config=$RAIDCONF
 	rc=$?
 	if [ $rc -ne 0 ]; then
 		ocf_exit_reason "Failed to set $mddev readonly (rc=$rc)"
@@ -246,7 +252,7 @@
 	rm -f $tmp_block_file
 	ocf_log info "block device file $1 missing, creating one in order to stop the array"
 	mknod $tmp_block_file b 9 $n
-	$MDADM --stop $tmp_block_file --config=$RAIDCONF
+	ocf_run $MDADM --stop $tmp_block_file --config=$RAIDCONF
 	rc=$?
 	rm -f $tmp_block_file
 	return $rc
@@ -254,12 +260,12 @@
 raid1_stop_one() {
 	ocf_log info "Stopping array $1"
 	if [ -b "$1" ]; then
-		$MDADM --stop $1 --config=$RAIDCONF &&
+		ocf_run $MDADM --stop $1 --config=$RAIDCONF &&
 			return
 	else
 		# newer mdadm releases can stop arrays when given the
 		# basename; try that first
-		$MDADM --stop `basename $1` --config=$RAIDCONF &&
+		ocf_run $MDADM --stop `basename $1` --config=$RAIDCONF &&
 			return
 		# otherwise create a block device file
 		mknod_raid1_stop $1
@@ -375,7 +381,7 @@
 		return $OCF_NOT_RUNNING
 	fi
 	if [ $HAVE_RAIDTOOLS != "true" ]; then
-		$MDADM --detail --test $mddev >/dev/null 2>&1 ; rc=$?
+		ocf_run $MDADM --detail --test $mddev >/dev/null 2>&1 ; rc=$?
 		case $rc in
 		0)	;;
 		1)	ocf_log warn "$mddev has at least one failed device."
@@ -403,9 +409,9 @@
 	if [ "$__OCF_ACTION" = "monitor" -a "$OCF_RESKEY_CRM_meta_interval" != 0 \
 		-a $TRY_READD -eq 1 -a $OCF_CHECK_LEVEL -gt 0 ]; then
 		ocf_log info "Attempting recovery sequence to re-add devices on $mddev:"
-		$MDADM $mddev --fail detached
-		$MDADM $mddev --remove failed
-		$MDADM $mddev --re-add missing
+		ocf_run $MDADM $mddev --fail detached
+		ocf_run $MDADM $mddev --remove failed
+		ocf_run $MDADM $mddev --re-add missing
 		# TODO: At this stage, there's nothing to actually do
 		# here. Either this worked or it did not.
 	fi
--- a/heartbeat/VirtualDomain	2021-09-14 07:56:25.201636489 -0400
+++ b/heartbeat/VirtualDomain	2021-09-19 11:11:06.856157215 -0400
@@ -460,7 +460,7 @@
 		try=$(($try + 1 ))
 		status=$(LANG=C virsh $VIRSH_OPTIONS domstate $DOMAIN_NAME 2>&1 | tr 'A-Z' 'a-z')
 		case "$status" in
-			*"error:"*"domain not found"|*"error:"*"failed to get domain"*|"shut off")
+			*"error:"*"domain not found"*|*"error:"*"failed to get domain"*|"shut off")
 				# shut off: domain is defined, but not started, will not happen if
 				#   domain is created but not defined
 				# "Domain not found" or "failed to get domain": domain is not defined
