--- a/heartbeat/Raid1	2016-02-03 05:57:55.000000000 -0500
+++ b/heartbeat/Raid1	2016-11-03 14:39:57.349612898 -0400
@@ -353,14 +353,18 @@
 #
 raid1_monitor_one() {
 	local mddev=$1
-	local md=`echo $mddev | sed 's,/dev/,,'`
+	if [ -e "$mddev" ]; then
+		local md=`readlink -f $mddev | sed 's,/dev/,,'`
+	else
+		local md=`echo $mddev | sed 's,/dev/,,'`
+	fi
 	local rc
 	local TRY_READD=0
 	local pbsize
 	# check if the md device exists first
 	# but not if we are in the stop operation
 	# device existence is important only for the running arrays
-	if [ "$__OCF_ACTION" != "stop" -a ! -b $mddev ]; then
+	if [ "$__OCF_ACTION" != "stop" -a ! -e $mddev ]; then
 		ocf_log info "$mddev is not a block device"
 		return $OCF_NOT_RUNNING
 	fi
