--- a/iscsi-scst/kernel/iscsi.c	2021-10-26 09:03:22.000000000 -0400
+++ b/iscsi-scst/kernel/iscsi.c	2022-04-28 09:27:47.533213983 -0400
@@ -2450,8 +2450,14 @@
 			    req_hdr->cmd_sn)) {
 			*status = ISCSI_RESPONSE_FUNCTION_COMPLETE;
 			res = 0;
-		} else
-			*status = ISCSI_RESPONSE_UNKNOWN_TASK;
+		} else {
+			PRINT_WARNING("Aborting out-of-window command ref_cmd_sn=%d (cmd_sn - 2048)=%d cmd_sn=%d",
+				      req_hdr->ref_cmd_sn,
+				      req_hdr->cmd_sn - 2048,
+				      req_hdr->cmd_sn);
+			*status = ISCSI_RESPONSE_FUNCTION_COMPLETE;
+			res = 0;
+		}
 	}
 
 out:
--- a/scst/src/dev_handlers/scst_vdisk.c	2022-12-22 22:08:08.000000000 -0500
+++ b/scst/src/dev_handlers/scst_vdisk.c	2022-12-23 12:10:03.063079572 -0500
@@ -3407,8 +3407,9 @@
 			 */
 			PRINT_WARNING("Closed FD on exec. Not active ALUA state "
 				"or not blocked dev before ALUA state change? "
-				"(cmd %p, op %s, dev %s)", cmd, cmd->op_name,
-				cmd->dev->virt_name);
+				"(cmd %p, op %s, dev %s, initiator %s)", cmd,
+				cmd->op_name, cmd->dev->virt_name,
+				cmd->sess->initiator_name);
 			scst_set_cmd_error(cmd, SCST_LOAD_SENSE(scst_sense_no_medium));
 			goto err;
 		}
