--- a/include/linux/types.h	2013-02-18 09:05:39.369970988 -0500
+++ b/include/linux/types.h	2013-02-18 09:09:48.705651890 -0500
@@ -24,6 +24,9 @@
 #define __bitwise
 #endif
 
+/* This ifndef is required by the xfsprogs package */
+#ifndef NO_ENDIAN_DEP_TYPES
+#define NO_ENDIAN_DEP_TYPES
 typedef __u16 __bitwise __le16;
 typedef __u16 __bitwise __be16;
 typedef __u32 __bitwise __le32;
@@ -33,6 +36,7 @@
 
 typedef __u16 __bitwise __sum16;
 typedef __u32 __bitwise __wsum;
+#endif /* NO_ENDIAN_DEP_TYPES */
 
 /*
  * aligned_u64 should be used in defining kernel<->userspace ABIs to avoid
