--- a/caching/hint_array.h	2013-12-02 09:07:53.405445903 -0500
+++ b/caching/hint_array.h	2013-12-02 09:08:11.337320887 -0500
@@ -56,7 +56,7 @@
 	class hint_array {
 	public:
 		typedef boost::shared_ptr<hint_array> ptr;
-		typedef typename persistent_data::transaction_manager::ptr tm_ptr;
+		typedef persistent_data::transaction_manager::ptr tm_ptr;
 
 		hint_array(tm_ptr tm, unsigned width);
 		hint_array(tm_ptr tm, unsigned width, block_address root, unsigned nr_entries);

--- a/caching/hint_array.cc	2013-12-02 09:08:00.628381553 -0500
+++ b/caching/hint_array.cc	2013-12-02 09:08:31.671509278 -0500
@@ -239,7 +239,7 @@
 {
 }
 
-hint_array::hint_array(typename hint_array::tm_ptr tm, unsigned width,
+hint_array::hint_array(hint_array::tm_ptr tm, unsigned width,
 		       block_address root, unsigned nr_entries)
 	: width_(check_width(width)),
 	  impl_(mk_array(tm, width, root, nr_entries))

--- a/caching/restore_emitter.cc	2013-12-02 09:12:19.550380402 -0500
+++ b/caching/restore_emitter.cc	2013-12-02 09:12:30.410276591 -0500
@@ -58,7 +58,7 @@
 		virtual void mapping(pd::block_address cblock,
 				     pd::block_address oblock,
 				     bool dirty) {
-			typename caching::mapping m;
+			caching::mapping m;
 			m.oblock_ = oblock;
 			m.flags_ = M_VALID;
 

--- a/persistent-data/data-structures/bitset.h	2013-12-02 09:04:32.309919971 -0500
+++ b/persistent-data/data-structures/bitset.h	2013-12-02 09:06:51.819047591 -0500
@@ -49,7 +49,7 @@
 	class bitset {
 	public:
 		typedef boost::shared_ptr<bitset> ptr;
-		typedef typename persistent_data::transaction_manager::ptr tm_ptr;
+		typedef persistent_data::transaction_manager::ptr tm_ptr;
 
 		bitset(tm_ptr tm);
 		bitset(tm_ptr tm, block_address root, unsigned nr_bits);

--- a/persistent-data/data-structures/bitset.cc	2013-12-02 09:04:39.301921623 -0500
+++ b/persistent-data/data-structures/bitset.cc	2013-12-02 09:06:38.875746849 -0500
@@ -30,7 +30,7 @@
 		class bitset_impl {
 		public:
 			typedef boost::shared_ptr<bitset_impl> ptr;
-			typedef typename persistent_data::transaction_manager::ptr tm_ptr;
+			typedef persistent_data::transaction_manager::ptr tm_ptr;
 
 			bitset_impl(tm_ptr tm)
 			: nr_bits_(0),

